### DISCUSSION TITLE
Push to arrays

### DISCUSSION THREAD

**Discussion by Rottweiler** [dotnet/csharplang#647] [General]

**Comment by Rottweiler** [comment]
Sorry pushed enter on my phone too quickly
But basically; implement pushing to a managed array like in PHP;

int data = new int [1,2,3]
data[] = 4

So that data is now 1,2,3,4

**Comment by HaloFour** [comment]
Considering the difference between arrays in C# and "arrays" in PHP I don't think that it's appropriate to enshrine such an operation in a special language operator.  Appending to an array is quite expensive and shouldn't be encouraged.

If it's something that you really want to do you can already define an extension method that would handle creation of a new array, appending the item(s) and returning the new value:

```cs
public static class ArrayExtensions {
    public static T[] Append<T>(this T[] array, T item) {
        if (array == null) throw new ArgumentNullException(nameof(array));
        T[] copy = new T[array.Length + 1];
        Array.Copy(array, copy, array.Length);
        copy[array.Length] = item;
        return copy;
    }
}
```

**Comment by jnm2** [comment]
@Rottweiler Arrays have fixed-size semantics everywhere they are used. People's code takes dependencies on the fact that nothing can cause them to resize.  `List<>` covers the cases where you need variable length, and it does it much more efficiently than resizing an array each time.

**Comment by YaakovDavis** [comment]
This is a weird syntax even for PHP.

**Comment by bondsbw** [comment]
I could maybe see doing something similar to this if there were a C# language mechanism for defining a `List<T>`.
```c#
var data = new list { 1, 2, 3 };  // equivalent to new List<int> { 1, 2, 3 }
data += 4;  // appends 4
```

_EDIT: Actually I guess that doesn't matter much, `+=` could be defined for `List<T>` regardless of declaration syntax._

**Comment by chergey** [comment]
There is much about arrays (i.e. fixed size, compiler optimizations, primitiveness) that make it unlikely their further development. But for collections (lists, dictionaries), such a feature would be nice.

**Comment by chergey** [comment]
@bondsbw 
Many features could be simulated but some of them better be hardcoded into the language (i.e. compiler), imho.

**Comment by CyrusNajmabadi** [comment]
There's almost no chance we would provide this.  However this may fall out if we ever allow extension operators.  Then you could just define your own ```+``` operator for lists.

**Comment by jnm2** [comment]
Also if we get object initializer syntax on already-created objects, that would look very similar.

**Comment by Joe4evr** [comment]
@jnm2 Soooo.... Essentially the "port VB's `With`" proposal again. 😝

**Comment by jnm2** [comment]
@Joe4evr It's around here somewhere.
